canvas/biplot.py

```diff
@@ -0,0 +1,499 @@
+                             'not %r.' % type(feature_df_metadata).__name__)
+
+    if sample_color_dict is not None:
+        if not isinstance(sample_color_dict, dict):
+            raise ValueError('`sample_color_dict` must be a `dictionary`, '
+                             'not %r.' % type(sample_color_dict).__name__)
+
+    if feature_color_dict is not None:
+        if not isinstance(feature_color_dict, dict):
+            raise ValueError('`feature_color_dict` must be a `dictionary`, '
+                             'not %r.' % type(feature_color_dict).__name__)
+
+    if sample_df_metadata is not None and sample_color_dict is None:
+        if sample_color_category is None:
+            raise ValueError('sample_color_category must be a specified')
+
+    if sample_df_metadata is not None and sample_color_dict is not None:
+        if sample_color_category is None:
+            raise ValueError('sample_color_category must be a specified')
+
+    if feature_df_metadata is not None and feature_color_dict is not None:
+        if feature_color_category is None:
+            raise ValueError('feature_color_category must be a specified')
+
+    if sample_drop_list is not None:
+        if not isinstance(sample_drop_list, list):
+            raise ValueError('`sample_drop_list` must be a `list`, '
+                             'not %r.' % type(sample_drop_list).__name__)
+
+    if sample_drop_col is not None:
+        if not isinstance(sample_drop_col, list):
+            raise ValueError('`sample_drop_col` must be a `list`, '
+                             'not %r.' % type(sample_drop_col).__name__)
+
+    if sample_df_metadata is not None:
+        if (sample_df.index != sample_df_metadata.index).any():
+            sample_df = sample_df.sort_index(axis=0)
+            sample_df_metadata = sample_df_metadata.sort_index(axis=0)
+
+    fig = plt.figure(figsize=figure_size)
+    ax = fig.add_subplot(111)
+
+    sample_colors = cm = plt.get_cmap(samp_col)
+    feature_colors = cm = plt.get_cmap(feat_col)
+
+    sample_group_append = []
+    colorVal = []
+
+    if sample_df_metadata is None:
+        print("Sample Block 1")
+        ax.plot(np.ravel(sample_df[sample_df_x]),
+                np.ravel(sample_df[sample_df_x]),
+                marker=samp_marker, linestyle='',
+                ms=samp_ms, alpha=samp_alpha)
+
+    if sample_df_metadata is not None and sample_color_dict is None:
+        print("Sample Block 2")
+        sample_df_copy = sample_df
+        sample_df_metadata_copy = sample_df_metadata
+
+        if len(sample_drop_list) > 0:
+            index_drop = sample_df_metadata[sample_color_category]
+            .apply(lambda x: x not in sample_drop_list)
+            sample_df_metadata = sample_df_metadata.loc[index_drop]
+            sample_df = sample_df.loc[sample_df_metadata.index]
+
+            for name, group in
+            sample_df.groupby(sample_df_metadata[sample_color_category]):
```

Comment:
Can we replace this with
```python
sample_groups = sample_df.groupby(sample_df_metadata[sample_color_category])
for name, group in sample_groups:
...
```

And replace all other instances of `sample_df.groupby(sample_df_metadata[sample_color_category])` with sample_groups?